### PR TITLE
Potions initial implementation

### PR DESCRIPTION
**Pull request by ajrb**
- All classic potions can be mixed.
- All classic potions can be drunk and the effects work.
- Failed potions are not created and a new message is displayed. (change from classic)
- We need to decide whether to restrict random potions in loot. Should they be restricted to classic ones only? I have a function for all and another for classic only - one could be used for mixing and purchasing, with the classic one used for loot. Opinions?
- Also added removeOne function to item collection which decrements stack and only removed item object when reaches 0. Please use it going forward where appropriate. Ta.
- Need to discuss mechanism for mult-effect potions @Interkarma - see the review comments.

### CODE REVIEW

File: Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs [hunk 1]
```diff
@@ -19,25 +19,41 @@ namespace DaggerfallWorkshop.Game.MagicAndEffects.MagicEffects
     /// </summary>
     public class InvisibilityNormal : ConcealmentEffect
     {
+        public static readonly string PublicKey = "Invisibility-Normal";
+
         public override void SetProperties()
         {
-            properties.Key = "Invisibility-Normal";
```
Comment (ajrb) [inline Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs]: Here's an example of making the EntityEffect key publicly available to support multi-effect potions while still keeping all potion recipes in EntityEffect classes as you wanted. We should discuss if there's a better way of doing this.

File: Assets/Scripts/Game/MagicAndEffects/Effects/Restoration/HealSpellPoints.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            properties.SubGroupName = TextManager.Instance.GetText(textDatabase, "spellPoints");
+            //properties.SpellMakerDescription = DaggerfallUnity.Instance.TextProvider.GetRSCTokens(1549);
+            //properties.SpellBookDescription = DaggerfallUnity.Instance.TextProvider.GetRSCTokens(1249);
+            properties.SupportMagnitude = true;
+            properties.AllowedTargets = EntityEffectBroker.TargetFlags_Self;
+            properties.AllowedElements = EntityEffectBroker.ElementFlags_MagicOnly;
+            properties.AllowedCraftingStations = MagicCraftingStations.PotionMaker;
+            //properties.MagicSkill = DFCareer.MagicSkills.Restoration;
+            //properties.MagnitudeCosts = MakeEffectCosts(8, 28);
+        }
```
Comment (ajrb) [inline Assets/Scripts/Game/MagicAndEffects/Effects/Restoration/HealSpellPoints.cs]: Please verify that I have these settings right for an effect that's only used by a potion recipe.

File: Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs [hunk 3]
```diff
@@ -19,25 +19,41 @@ namespace DaggerfallWorkshop.Game.MagicAndEffects.MagicEffects
...
+            properties.Key = PublicKey;
```
Comment (ajrb) [inline Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs]: Here's an example of making the EntityEffect key publicly available to support multi-effect potions while still keeping all potion recipes in EntityEffect classes as you wanted. We should discuss if there's a better way of doing this.

**Comment by Interkarma** [comment]
Thank you for all your hard work on this mate! I'm starting to work through it now alongside a few other items. Cheers!

File: Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs (hunk 3 above)
Comment (Interkarma) [reply Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs]: I'm happy with this for now. Thank you for overcoming this problem. Let's see how it goes before we worry about refining. 👍

File: Assets/Scripts/Game/MagicAndEffects/Effects/Restoration/HealSpellPoints.cs (hunk 2 above)
Comment (Interkarma) [reply Assets/Scripts/Game/MagicAndEffects/Effects/Restoration/HealSpellPoints.cs]: This looks OK on a read. Assuming tests OK in game as well? I'll test soon.

File: Assets/Scripts/Game/UserInterfaceWindows/DaggerfallPotionMakerWindow.cs [hunk 4]
```diff
@@ -238,7 +240,26 @@ void RemoveFromCauldron(DaggerfallUnityItem item)
...
+            //IEntityEffect potionEffect = GameManager.Instance.EntityEffectBroker.GetPotionRecipeEffect(recipe);
+            PotionRecipe potionRecipe = GameManager.Instance.EntityEffectBroker.GetPotionRecipe(recipeKey);
+            if (potionRecipe != null)
+            {
+                Debug.LogFormat("Potion matched: {0}", potionRecipe.DisplayName);
+                GameManager.Instance.PlayerEntity.Items.AddItem(ItemBuilder.CreatePotion(recipeKey));
+            }
+            else
+            {
+                // Changed from classic, don't create useless 'Unknown Powers' potions.
```
[the remark this answers is not in the document]
Comment (Interkarma) [reply Assets/Scripts/Game/UserInterfaceWindows/DaggerfallPotionMakerWindow.cs]: I feel the same, I barely touched potions in classic to be honest.

File: Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs (hunk 3 above)
Comment (ajrb) [reply Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs]: Okay should I change all of the entity effect classes to follow the same pattern then? Think I will change the name from PublicKey to EffectKey though.

Comment (Interkarma) [reply Assets/Scripts/Game/MagicAndEffects/Effects/Illusion/InvisibilityNormal.cs]: I'm happy for that, if you'd like to work through it. I'll make sure I follow the same pattern for any future effects that I write.

BTW, now you're across the effect system would you be interested in helping to build any of the remaining spell effects, item effects, advantages/disadvantages, or artifact effects? I plan to start knocking these over again soon once I'm done having fun with vampirism.

**Pull request merged**